### ISSUE TITLE
print propositions

### ISSUE THREAD

**Issue by benbrastmckie** [benbrastmckie/ModelChecker#54]
The print implementation is working great! I had a few ideas for how it can be improved and made more modular to accommodate the addition of new operators. I am also motivated to make the printing for each operator defined alongside the semantics for that operator so that users can introduce both together, since changes to one may require changing the other when developing new semantic clauses. Here are some thoughts for how this can be refactored.

Currently `rec_print` is called in `print_inputs_recursively`, and `print_proposition` is a method of `Proposition` which gets called for each `sentence_object` in `rec_print`. In the special case where the `prefix_operator` does not have a `print_operator` method (all besides the CF operator), then the arguments are printed. I'm thinking this can be refactored as follows:

1. Moving `rec_print` into `print_proposition` in `Proposition` which calls a `print_proposition` method defined in the `main_operator` for that `sentence_object` if `self.arguments` is truthy, and otherwise prints the sentence letter. This way of dispatching to different operators should mirror the way `true_at` and `false_at` work in `Semantics`.
2. The `print_proposition` method defined in each operator will only define how a sentence with that operator as the `prefix_operator` will print, calling the `print_proposition` method in each of its `sentence_object` arguments for recursion.

The hope here is that this will mirror how the semantics works by bouncing around between operators. I was also thinking about an alternative which I'm still uncertain about. The idea would be to move `Sentence` to `exposed_things.py` and include a method called `print_sentence` which would replace what `rec_print` is doing now and could be called from any `sentence_object`. When `print_sentence` is called, all it does is look up the `prefix_operator` (if any) and calls `print_proposition` in that operator, and otherwise calls `print_proposition` in `Proposition` to print the sentence letter. That way `print_proposition` in `Proposition` would only be responsible for printing sentence letters (not a big advantage). Additionally, we wouldn't need to look into the `proposition` attribute for each `sentence_object` in order to state the recursive call (this might be slightly more intelligible to users, but not a big advantage). Lastly, moving `Sentence` to `exposed_things.py` could help make some of the key attributes more intelligible to users, and allow them to adapt the definition (also not a big advantage). Against these considerations, exposing `Sentence` adds complexity to what a user will have to face to build their semantics. Also, I was thinking `exposed_things.py` could eventually be renamed `semantic.py` and adding `Sentence` would make that less appropriate. All in all, I'm not convinced that this is the best way to go, but thought I'd document the idea here. Curious to hear your thoughts.
`

**Comment by mbuit82** [comment]
I'm not sure how many more operators need specific printing methods besides CFs, and those that do (or would in other people's semantics, say we encounter something like that in Champollion's) seem to be on an _ad hoc_ basis. So I think the current printing is fine, since it is pretty hidden in terms of mechanics. And I think the way it is currently set up is that it'll do `rec_print` as a default (ie printing vers and fals for L and R args of a binary operator), except if the main operator has a `print_operator` method, in which case it never does `rec_print`/default things. 

I also think keeping `Sentence` hidden is good because it feels like an internal tool to transmit information from different (often hidden in any case) parts of the program. 

As a related issue, I think the general size of `exposed_things` may be a little concerning at the moment anyways at about 1000 lines of code (granting a lot of them are Z3 constraints or comments at the moment or operators); if users need to write ~1000 lines of code to check their own semantics or look in a 1000 line file to make tweaks to an existing one the value in ease of replication is not as strong. Granted, a big reason `exposed_things` is currently so long is because of all the operators, which are about 600 lines of code, so a quick fix would be to make an `operators` module for packaging, and if users want to define more operators they'd have to live with making a long file, since it seems hard to shorten their definitions for genuinely shorter operators. But even at 400 lines of code that may feel daunting for users to write on their own; my general hunch is we should aim to make what users have to write for their own semantics as small as possible. (For reference to other academic/scientific python programs, you can implement a neural net in using `PyTorch` in less than 100 lines of code (with long docstrings and all—without those in principle you could do it in like 10-20 lines for a simple model); with helper functions for loading data and all maybe less than 200 total. And packages I used over my summer internship extensively, `cldf` and `lingpy`, require a typical user to write no more than 100-200 lines of code. I know these aren't completely comparable since users have to write long Z3 constraints in any scenario for `model_checker`, but it gets the idea across.) And because of that only adding to it with `Sentence` may not be worth it.

**Comment by benbrastmckie** [comment]
I think dividing `exposed_things.py` is a great idea. I'm happy to do that. This is also an opportunity to rename modules in a suggestive way, though `exposed_things.py` has served us well.

I also agree about continuing to leave `Sentence` hidden. That makes good sense, though probably some mention of it will occur in the documentation. Thinking about the docs has been a good way for me to get clearer about the code base in general and to see ways of refining the resulting methodology.

Along these lines, I still think it is important to have print methods for all operators. I have an idea in mind and can pursue this on a branch. I don't think it will be too involved as they are all pretty trivial with the exception of the CF operators and modals. My aim here is to avoid ad hoc exceptions. One way to think about it is that the product is not really an implementation but rather a methodology where the implementation is an instance of that methodology. So even if the current implementation is great as it is where it would make sense to make ad hoc additions, it is uniformity of the method that really inspires this idea. But let's see what I can come up with.

I think the DefinedOperator can be worked on separately from the print methods I want to add, though some minor tweaks may end up being required later.

**Comment by mbuit82** [comment]
I see, sounds good! DefinedOperator is working now. I tested ConditionalOperator, BiconditionalOperator, and DefNecessaryOperator (defining `\\Box A` as `(\\top \\boxright A)`). A few changes needed to be made here and there in a couple places (like `Semantics.extended_verify` and `extended_falsify`). I also added something that sorts the sentences in the `all_sentences` dictionary from least complex to most; this ended up being relevant for DefinedOperators. To this end I needed to make an `infix` helper function (identical to `Sentence.infix` except without the `self` argument) because I needed to use `infix` at one point without any `Sentence` instance yet. 

The heart of the implementation is in `DefinedOperator.find_verifiers_and_falsifiers` and `DefinedOperator.get_updated_Sentence`. The code is a bit confusing, but I think there are no problems. The biggest questionable thing was I had to make a couple of "artificial" `Sentence` objects in DefinedOperators when finding verifiers and falsifiers; these never are saved to the bigger `all_sentences` dictionary and I don't think this matters because of the recursive call of `DefinedOperator.get_updated_Sentence`.

**Comment by benbrastmckie** [comment]
Nice work! The changes you indicate are really helpful. I'll take a look.

As for `infix` I was thinking this is a useful enough function that it should belong to `hidden_helpers.py` which we could be renamed to `utils.py` and called on in the API to add certain general purpose functions. I think `prefix` is like this too.

Sorting seems like a good regardless. I had removed what I had added before since it wasn't needed, but not a bad thing to have anyways. I'll try to dig into the DefinedOperator stuff and expand the docsrings.

I was originally thinking the DefinedOperator would also specify a print method by drawing on the print methods for the operators from which it is defined, but even if we could get that to work, I'm not sure it is better than having the ability to specify the print method directly. For instance, it is nice that the defined modal can print the whole sentence and print the argument in each world instead of also printing top as it might do if it were to generate a print method from top and CF. So I think it is great how it is.

**Comment by mbuit82** [comment]
I agree, I was actually thinking about the `\\Box` case specifically—seems like the `DefinedOperator` really shouldn't have the print methods of its subparts, or at least not in all cases

**Comment by benbrastmckie** [comment]
> few changes needed to be made here and there in a couple places (like `Semantics.extended_verify` and `extended_falsify`).

I'm trying to figure out why the extra condition is needed but can seems to find an example where this condition is met. Here is what I'm trying to get to print:

```
    def extended_verify(self, state, prefix_object, eval_world):
        if isinstance(prefix_object, syntactic.Operator):
            # B: I don't think this ever gets called DISCUSS
            print("TEST CHANGE")
            return prefix_object.extended_verify(state, eval_world)
        if str(prefix_object[0]).isalnum():
            return self.verify(state, prefix_object[0])
        op, args = prefix_object[0], prefix_object[1:]
        return op.extended_verify(state, *args, eval_world)
    
    def extended_falsify(self, state, prefix_object, eval_world):
        if isinstance(prefix_object, syntactic.Operator):
            # B: I don't think this ever gets called DISCUSS
            print("TEST CHANGE")
            return prefix_object.extended_falsify(state, eval_world)
        if str(prefix_object[0]).isalnum():
            return self.falsify(state, prefix_object[0])
        op, args = prefix_object[0], prefix_object[1:]
        return op.extended_falsify(state, *args, eval_world)
```

**Comment by mbuit82** [comment]
Hmm, that seemed to be needed for derived definitions with zero-place operators (like the definition of `\\Box` in terms of CFs). Though it may be that it ended up not being needed. If there's no issues raised for those cases, then we can probably get rid of the condition

**Comment by benbrastmckie** [comment]
This is to follow up on the finding that the extra clauses were needed for `\\top`, but do not seem to be working. It would be good to find a way to remove these extra conditions. You mentioned that this might get easier after reworking `DefinedOperator` since the trouble seems to be a clash coming from the fact that not all defined prefix sentence objects are lists as they were before, but this may be irrelevant once the update functions in `Sentence` derive the `prefix_type` in terms of primitive operators.

**Comment by benbrastmckie** [comment]
I just moved the `calculate_alternative_worlds` method to `Semantics` is then called in the relevant operator classes and got rid of `calculate_true_worlds` which was a mistake. Instead modals should print over all worlds.

There are no other printing issues that I can see so will close this issue for now.
